### DISCUSSION TITLE
How to do aggregation on a tree structure?

### DISCUSSION THREAD

**Discussion by divvan** [ravendb/ravendb#19864] [General]
Part of my application handles project management. I model the work that needs to be done during a project using a hierarchical tree structure of tasks and sub-tasks. (This is similar to a "Work Breakdown Structure" in traditional, non-agile project management: https://en.wikipedia.org/wiki/Work_breakdown_structure ). The tree will have 4 levels, which I'll call A to D, with a one-to-many relationship between them.

    A--[1:N]--B--[1:N]--C--[1:N]--D

All the work in the project is done at the lowest level, D. At the start of the project, every item in D is assigned an EstimatedHours value. Then as each item's work is being done, the user increases its PercentComplete value from 0 up to 100. The PercentComplete and EstimatedHours are then used to calculate the item's CompletedHours.

These three values - EstimatedHours, PercentComplete and CompletedHours - are used at every level of the tree. They are entered at the D level, but then they need to be aggregated at the higher levels as well.

MapReduce seems to be the way to do aggregations in RavenDB, but it looks like it will only work at one level at a time. So, to do a multi-level aggregation, do I need to save the changes at the lowest level, use MapReduce to calculate the values at the next level up, save its changes, and then repeat this up to the top of the tree? Or is there an easier way to do it?

My main concern with doing repeated SaveChanges() is that if the application crashes partway through the process, the data will be inconsistent - the lower levels will have the latest values but the higher levels won't.

**Comment by karmeli87** [comment]
Hi, 

I'm not 100% sure that I follow what you are trying to achieve here. If you have a specific example (with some code & classes) that would great.

links that could be relevant to you:
https://ravendb.net/docs/article-page/6.2/csharp/indexes/indexing-hierarchical-data
https://ravendb.net/docs/article-page/6.2/csharp/indexes/indexing-polymorphic-data

**Comment by ayende** [comment]
That is an interesting question. I wrote a blog post (currently in draft form) to answer it.
You can find it here:

https://ayende.com/blog/201861-C/aggregating-trees-with-ravendb?key=817db81f86904471a09d2fd4ef094fff

Would love to hear your feedback

**Reply by divvan** [reply]
That's great, Oren! I'm amazed how versatile RavenDB is.

**Comment by ayende** [comment]
Here is the actual blog post for this:

https://ayende.com/blog/201861-C/aggregating-trees-with-ravendb?key=817db81f86904471a09d2fd4ef094fff

**Reply by mcrio** [reply]
Thanks Oren, great example. Being able to utilize JS is so powerful.
